source/_posts/2015-10-11-a-configuration-management-rosetta-stone-configuring-sensu-with-puppet.markdown

```diff
@@ -0,0 +1,352 @@
+  inconsequential. As long as Sensu behaved the same I considered it complete.
+* No considerations for security (out of scope for this exercise)
+
+## Review of Each Tool
+
+### Puppet
+
+#### Puppet In General
+
+Puppet is my "native language" when it comes to configuration management, so it
+is a little hard for me to imagine what it is like to *not* know what it is
+like to know exactly how it works.
+
+Puppet has a custom DSL to describe configuration in terms of "types". These are
+the primitives that you can build infrastructure upon, things like "files, "package",
+and "service". Third party modules can extend that language with custom types,
+allowing you to abstract over the "raw" types. For example, the RabbitMQ has a
+type for providing `rabbitmq_user`s, which do not correspond to a particular
+config file or anything, but instead can only be added by special invocations
+of the `rabbitmqctl` command.
+
+Puppet strongly emphasizes code-reuse. The [Puppet Forge](https://forge.puppetlabs.com/) is the registry where you can upload and
+share modules. The Forge has a number of methods to help divine code quality.
+It also exposes "officially supported" and "officially approved" modules, for
+extra approval stamps. While the forge may have a very "long tail" of modules
+that do very common tasks, the set of officially-supported and
+officially-approved modules leaves behind a good selection of high-quality
+modules ready for re-use.
+
+#### Review of the Sensu Puppet Module
+
+The [sensu-puppet module](https://github.com/sensu/sensu-puppet) is a
+first-class citizen in the Sensu world. It has native types for the Sensu JSON
+files that it manages, as well as a `sensu-gem` type for easily installing
+rubygems with the embedded Sensu ruby.
+
+The Sensu Puppet module only manages Sensu, and has no integration with any
+other RabbitMQ, Redis, or any other module. To me this is expected, in the
+Puppet world it would be the job of a `profile` to combine the Sensu module
+with RabbitMQ and other things. For the most part this integration is left as
+an exercise to the reader.
+
+The Sensu Puppet module also doesn't manage Uchiwa. That requires a
+[different puppet module](https://github.com/Yelp/puppet-uchiwa). Again to me
+this is a good thing, I hate it when tools try to do too much.
+
+The actual codebase is actively maintained and reasonably active, with a few
+releases per year. The Puppet Forge has almost perfect scores for it's quality
+ratings. The code has excellent unit test and acceptance test coverage. As far
+as puppet modules go, the Sensu puppet module is a great example of a
+well-maintained piece of code.
+
+One downside the "completeness" of the module is that sometimes new features of
+Sensu are released, and the puppet-module will lag. The configuration inputs to
+the puppet module are well-typed, and not just free-form hashes. This gives a
+lot of guard-rails and helps ensure config files are correct before they hit
```

Comment:
Hyphenating this word is not [popular nowadays](https://books.google.com/ngrams/graph?content=guard-rails%2Cguardrails%2Cguard+rails&year_start=1800&year_end=2000&corpus=15&smoothing=3&share=&direct_url=t1%3B%2Cguard%20-%20rails%3B%2Cc0%3B.t1%3B%2Cguardrails%3B%2Cc0%3B.t1%3B%2Cguard%20rails%3B%2Cc0)